Ping check

lrettig opened the pull request:
Closes #718 

WIP, still needs tests.

When successfully pinging a peer (pulled randomly out of the routing table/addrbook), mark the peer as having been successfully pinged.
When receiving a ping from a peer that's not already marked as having been successfully pinged, send a ping to their advertised addr/port.
Refuse to respond to GetAddresses from a peer that hasn't recently passed this ping check.

lrettig committed:
Begin to lay groundwork for ping check

Extend KnownAddress with a lastpong time. Extend the address book with a
map of pinged addresses. Add a comment where the code needs to go in
verifyPinger.

lrettig committed:
Initial support for ping checks

Extend KnownAddress with a lastping property, and getter/setter.
Update Ping to set lastping upon successful ping, and to send a ping
when necessary.
Update serialization and interfaces as necessary.

lrettig committed:
Perform ping check before sending addresses

lrettig committed:
Don't attempt to ping in newGetAddresses

On second thought, assume that the ping was already sent, and just fail
if it didn't succeed.

lrettig committed:
Minor code cleanup

lrettig commented:
This isn't quite ready for merge yet (needs testing), but @y0sher if you have a moment I'd appreciate your feedback here - does this look reasonable? Thanks!

y0sher commented:
Nice work! looks good to me. please proceed with tests :)

lrettig committed:
Fix spacing

lrettig committed:
Add a mocked function to make mockAddrBook happy

p2p/discovery tests now passing locally.

lrettig committed:
Replace foo() named coroutine with IIFE

Immediately invokved function express is cleaner and less verbose

lrettig committed:
Add tests for KnownAddress NeedsPing

lrettig committed:
Add basic test for verifypinger

Call the method twice. The first time it should fail since the address
book is empty. After initialization, the second call should pass.

Add an error return to the method if address lookup fails, to prevent
null pointer exception.

lrettig committed:
Add placeholder comment for outstanding todo item

lrettig committed:
Some minor cleanup

lrettig committed:
Formatting, make linter happy

lrettig commented:
@y0sher Added some basic tests for `KnownAddress` changes and for `verifyPinger`. I haven't figured out how to test that the ping is actually sent (or received)--it seems that this would require some complex mocking. Tests are passing locally but for some reason they appear to be timing out in Travis, I'm not sure if this is a known issue.

y0sher commented:
@lrettig Ok so, you've missed some important info. `Ping` protocol is written using a `MessageServer` and a `MessageServer` uses a demuxer (either `swarm` or `udpmux` which eventually should be merged). this demuxer exposes the `SendMessage` which lets consumers of the P2P api (which discovery is one of them through `MessageServer`) to send messages using only p2p public keys. this means there's a backing data structure which map public keys to IP addresses (the `Addrbook`). 
So the only special case where this logic fails is on `Ping`. because we want to send messages to the pinger before we have it in the addrbook, in the current impl there's probably no way to do it without adding him to the addrbook beforehand. so a working (but kinda hacky) impl could add the address and then remove it if we've failed to ping it. this still solves the reflected DoS attack since you won't send the addresses if you didn't get the ping back. 

a working impl might look like that : 
```go

func (p *protocol) verifyPinger(from net.Addr, pi *node.NodeInfo) error {
	// todo : Validate ToAddr or drop it.
	// todo: decide on best way to know our ext address

	if err := pi.Valid(); err != nil {
		return err
	}

	// Check the address provided by pinging it (if we haven't already, recently).
	// This helps prevent reflective DoS attacks.
	ka, err := p.table.LookupKnownAddress(pi.PublicKey())
	if err != nil {
		ipfrom, _, _ := net.SplitHostPort(from.String())
		pi.IP = net.ParseIP(ipfrom)
		p.table.AddAddress(pi, pi)
		ka, _ = p.table.LookupKnownAddress(pi.PublicKey())
	}

	if ka.NeedsPing() {
		peer := ka.na.PublicKey()
		// Send the new Ping in a coroutine so we first respond to the incoming Ping
		go func() {
			if err := p.Ping(peer); err != nil {
				// All we can do here is print a warning. We've already responded with a pong,
				// and the peer will not be added to the pingable list.
				p.logger.Warning("Failed response to ping to Peer: %v", peer.String())
				p.table.RemoveAddress(pi.PublicKey())
			}
		}()
	}
	return nil
}
```

though, an implementation where we add an option to send a message without having the addr in the addrbook (using the ip addr itself and not the public key). would be a nice feature for this usage and will increase security. it might have a big overhead if there's a way to solve it without this fundamental change.

y0sher reviewed (changes requested):
I've commented about some possible ways to finish this pr

lrettig committed:
Add new addr to addrbook before attempting lookup

Per Yosher's suggestions.

Also reimplement Lookup using LookupKnownAddress to remove a little bit of code duplication

lrettig commented:
Thanks for the suggestions, I'll update the PR with this shortly.

> so a working (but kinda hacky) impl could add the address and then remove it if we've failed to ping it. this still solves the reflected DoS attack since you won't send the addresses if you didn't get the ping back.

Is there any chance of a race condition here where a peer sends an initial `PING`, and then immediately follows up with `GET_ADDRESSES` before the initial `PING` has failed and the address has been dropped from the addrbook?

> 		ka, _ = p.table.LookupKnownAddress(pi.PublicKey())

Is there any chance that this second lookup could ever fail? We're essentially asserting here that it cannot.

lrettig committed:
Retab

:facepalm:

y0sher commented:
cool, nice progress! I tried running it on my machine and it fails as well, it seems that the problem is some how related to the race you described. maybe your machine is faster so the race never happens.

the sender sends ping and gets pong right away so he sends get addresses though the receiver haven't finished his ping(and haven't received pong) yet.

So as I see this there are a few options to solve this race, 
- we would try and ping the sender whenever we received a `GetAddresses` request if we found out this node `NeedsPing`.
- the sender would check whether he got pinged from the "sendee" lately (note: this will probably require more state, similar to `NeedsPing` but more like do I need ping from him ?) though with this solution we'll still be in kinda the same problem. (we'll have to wait for the ping-pong to end somehow)

I'd appreciate your input though both solutions seem valid to me.

lrettig commented:
I'm stuck on these tests. Travis reports that `github.com/spacemeshos/go-spacemesh/hare/eligibility` fails, then it hangs on `github.com/spacemeshos/go-spacemesh/p2p/discovery`. The first is passing for me locally. The second hangs for me forever, but also on the develop branch (#1677).

lrettig committed:
Improve debugging

lrettig committed:
Perform pingcheck in response to getaddresses

Rather than trying to perform it in response to an incoming ping, which
causes problems. It's better to go ahead and (tentatively) add the peer
(which sent the ping) to the address book, then verify it with a
pingcheck later, when it sends getaddresses. Otherwise we have a race
condition where the getaddresses is received before the pingcheck
occurs.

lrettig commented:
I think there are a few options to solve the race:

- Respond to an incoming `PING` with a _synchronous_ outgoing `PING` to verify the address advertised by the sender (my previous attempt sent the outgoing ping async, which meant that the sender would immediately send `GETADDRESSES` before it had received the incoming `PING`, leading to the race condition). As you highlighted this would require more state as the sender of the initial `PING` would have to remember that it was waiting for a `PONG` and allow itself to respond to an incoming `PING` from the peer it was trying to handshake with. This is messy. I'd rather PING/PONG messages be stateless and dumb.
- (As you suggested) When an initial `PING` is received, go ahead and tentatively add the sender to the address book, but don't actually respond to an incoming `GETADDRESSES` request from that peer until a ping check has been performed successfully. If the ping check fails, drop the peer from the address book.
- I think the most robust way to solve this would be for the addrbook itself to be a "service" that can receive a request to add an address. When it sees a new address, it goes off on its own and performs a ping check to validate the address; if it doesn't hear back within a specified period of time, it expires that request and doesn't actually update the addrbook. However, this would be much more complex (and in practice is nearly identical to the second).

I implemented the second in my latest code changes.

lrettig committed:
Minor cleanup of last commit

lrettig committed:
Update ping-related tests

Instrument various known address lookup functions to return the right
thing, or else to return something that looks like we just pinged it.

lrettig committed:
Minor cleanup

Remove some redundant code from last commit, improve a debug message

lrettig committed:
Run p2p.discovery tests in verbose mode

Try to figure out why travis is hanging

y0sher committed:
try giving travis bigger timeout

lrettig committed:
Improve some debug output

antonlerner commented:
closed due to inactivity

lrettig commented:
This is superseded by #2294
